pkg/controller/issueapi/batch_issueapi.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/observability"
+)
+
+// HandleBatchIssue shows the page for batch-issuing codes.
+func (c *Controller) HandleBatchIssue() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if c.config.IsMaintenanceMode() {
+			c.h.RenderJSON(w, http.StatusTooManyRequests,
+				api.Errorf("server is read-only for maintenance").WithCode(api.ErrMaintenanceMode))
+			return
+		}
+
+		resp := &api.BatchIssueCodeResponse{}
+
+		ctx := r.Context()
+		realm := controller.RealmFromContext(ctx)
+		if !realm.AllowBulkUpload {
+			controller.Unauthorized(w, r, c.h)
+			return
+		}
+
+		hasSMSConfig, err := realm.HasSMSConfig(c.db)
+		if err != nil || !hasSMSConfig {
+			controller.InternalError(w, r, c.h, err)
+			return
+		}
+
+		result := &issueResult{
+			httpCode:  http.StatusOK,
+			obsBlame:  observability.BlameNone,
+			obsResult: observability.ResultOK(),
+		}
+		defer observability.RecordLatency(&ctx, time.Now(), mLatencyMs, &result.obsBlame, &result.obsResult)
+
+		var request api.BatchIssueCodeRequest
+		if err := controller.BindJSON(w, r, &request); err != nil {
+			c.h.RenderJSON(w, http.StatusBadRequest, api.Error(err))
+			result.obsBlame, result.obsResult = observability.BlameClient, observability.ResultError("FAILED_TO_PARSE_JSON_REQUEST")
+			return
+		}
+
+		authApp, user, err := c.getAuthorizationFromContext(r)
+		if err != nil {
+			c.h.RenderJSON(w, http.StatusUnauthorized, api.Error(err))
+			result.obsBlame, result.obsResult = observability.BlameClient, observability.ResultError("MISSING_AUTHORIZED_APP")
+			return
+		}
+
+		if authApp != nil {
+			realm, err = authApp.Realm(c.db)
+			if err != nil {
+				c.h.RenderJSON(w, http.StatusUnauthorized, nil)
+				result.obsBlame, result.obsResult = observability.BlameClient, observability.ResultError("UNAUTHORIZED")
+				return
+			}
+		} else {
+			// if it's a user logged in, we can pull realm from the context.
+			realm = controller.RealmFromContext(ctx)
+		}
+		if realm == nil {
+			c.h.RenderJSON(w, http.StatusBadRequest, api.Errorf("missing realm"))
+			result.obsBlame, result.obsResult = observability.BlameServer, observability.ResultError("MISSING_REALM")
+			return
+		}
+
+		// Add realm so that metrics are groupable on a per-realm basis.
+		ctx = observability.WithRealmID(ctx, realm.ID)
+
+		// TODO(whaught): batch size limits
+		logger := logging.FromContext(ctx).Named("issueapi.HandleBatchIssue")
+
+		resp.Codes = make([]*api.IssueCodeResponse, len(request.Codes))
+		for i, singleIssue := range request.Codes {
+			resp.Codes[i] = c.issue(ctx, singleIssue, realm, user, authApp, result)
+			if result.errorReturn != nil {
+				if result.httpCode == http.StatusInternalServerError {
+					controller.InternalError(w, r, c.h, errors.New(result.errorReturn.Error))
```

Comment:
Probably want a multi-error instead?

Reply:
this is what I asked in the API definition about failure semantics - we need to be able to know which which rows were successfully processed so they don't get retreied

they may have send real people SMS already

Reply:
added in the rest of the error handling